android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorage.java

```diff
@@ -1,55 +1,159 @@
 package com.oblador.keychain.cipherStorage;
 
 import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
 
 import com.oblador.keychain.SecurityLevel;
 import com.oblador.keychain.exceptions.CryptoFailedException;
 import com.oblador.keychain.exceptions.KeyStoreAccessException;
 
+import java.security.Key;
+
+@SuppressWarnings({"unused", "WeakerAccess"})
 public interface CipherStorage {
-    abstract class CipherResult<T> {
-        public final T username;
-        public final T password;
-
-        public CipherResult(T username, T password) {
-            this.username = username;
-            this.password = password;
-        }
-    }
+  //region Helper classes
 
-    class EncryptionResult extends CipherResult<byte[]> {
-        public CipherStorage cipherStorage;
+  /** basis for storing credentials in different data type formats. */
+  abstract class CipherResult<T> {
+    public final T username;
+    public final T password;
 
-        public EncryptionResult(byte[] username, byte[] password, CipherStorage cipherStorage) {
-            super(username, password);
-            this.cipherStorage = cipherStorage;
-        }
+    public CipherResult(T username, T password) {
+      this.username = username;
+      this.password = password;
     }
+  }
 
-    class DecryptionResult extends CipherResult<String> {
-      private SecurityLevel securityLevel;
-
-      public DecryptionResult(String username, String password, SecurityLevel level) {
-            super(username, password);
-            securityLevel = level;
-        }
+  /** Credentials in bytes array, often a result of encryption. */
+  class EncryptionResult extends CipherResult<byte[]> {
+    public final CipherStorage cipherStorage;
```

Comment:
storing the reference here perhaps not needed - the only place where it is used is in `PrefsStorage#storeEncryptedEntry` and `KeychainModule#migrateCipherStorage` but there, it's easy to provide it as a function parameter which seems a little cleaner to me.

Reply:
agree. It's not needed. Replacing it now by string value.